kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java

```diff
@@ -355,74 +346,150 @@ private Module resolveConfigBubbles(Definition def, Module inputModule, RuleGram
-                    .flatMap(b -> performParse(cache.getCache(), parser, realScanner, b))
-                    .map(this::upAlias)
-                .collect(Collections.toSet());
-
-            if (needNewScanner) {
-                realScanner.close();//required for Windows.
+            if (!replacedBubbles.isEmpty()) {
+                // make a new module with replaced bubbles
+                return Module(m.name(),
+                        m.imports(),
+                        stream((Set<Sentence>) m.localSentences().$minus$minus(immutable(replacedBubbles)).
+                                $bar(immutable(cachedParses))).collect(Collections.toSet()),
+                        m.att());
+            }
+            return m;
+        }, "").apply(defWithConfig);
+
+        // prepare scanners for remaining bubbles
+        // scanners can be reused so find the bottom modules which include all other modules
+        java.util.Set<Module> botMods = getBotModules(defWithCaches.modules()).stream().filter(m -> m.sentences().filter(s -> s instanceof Bubble).size() != 0).collect(Collectors.toSet());
+        // prefer modules that import the main module. This way we avoid using the main syntax module which could contain problematic syntax for rule parsing
+        java.util.Set<Module> orderedBotMods = new java.util.LinkedHashSet<>();
+        for (Module m : botMods) {
+            if (m.name().equals(defWithCaches.mainModule().name()) || m.importedModuleNames().contains(defWithCaches.mainModule().name()))
+                orderedBotMods.add(m);
+        }
+        orderedBotMods.addAll(botMods);
+
+        // map the module name to the scanner that it should use when parsing
+        java.util.Map<String, Module> donorModule = new HashMap<>();
+        for (Module m : mutable(defWithCaches.modules())) {
+            if (stream(m.localSentences()).anyMatch(s -> s instanceof Bubble)) {
+                Module scannerModule = orderedBotMods.stream().filter(bm -> m.equals(bm) || bm.importedModuleNames().contains(m.name())).findFirst()
+                        .orElseThrow(() -> new AssertionError("Expected at least one bottom module to have a suitable scanner: " + m.name()));
+                donorModule.put(m.name(), scannerModule);
             }
-
-            return Module(module.name(), module.imports(),
-                    stream((Set<Sentence>) module.localSentences().$bar(ruleSet).$bar(claimSet).$bar(contextSet).$bar(aliasSet)).filter(b -> !(b instanceof Bubble)).collect(Collections.toSet()), module.att());
         }
+        // create scanners
+        Map<Module, ParseInModule> donorScanners = new HashSet<>(donorModule.values()).parallelStream().map(x -> {
+            ParseInModule pim = RuleGrammarGenerator.getCombinedGrammar(gen.getRuleGrammar(x), isStrict, profileRules, files);
+            pim.getScanner(options.global);
+            return new Tuple2<>(x, pim);
+        }).collect(Collectors.toMap(Tuple2::_1, Tuple2::_2));
```

Comment:
Perhaps should be called `donorParseInModules`?

Reply:
Also, do we know at this point that every scanner we generate here is actually needed? We haven't actually checked (as far as I can tell) that each generate scanner here will be used.

Reply:
The previous step minimizes the number of scanners needed:
`// map the module name to the scanner that it should use when parsing`
`donorParseInModules` does sound better.